File: src/main/java/ru/arc/farm/Mine.java
package ru.arc.farm;

import ru.arc.ARC;
import ru.arc.configs.Config;
import ru.arc.configs.ConfigManager;
import ru.arc.util.CooldownManager;
import ru.arc.util.ParticleManager;
import ru.arc.util.TextUtil;
import ru.arc.util.Utils;
import com.destroystokyo.paper.ParticleBuilder;
import com.jeff_media.customblockdata.CustomBlockData;
import com.sk89q.worldedit.bukkit.BukkitAdapter;
import com.sk89q.worldedit.math.BlockVector3;
import com.sk89q.worldedit.regions.CuboidRegion;
import com.sk89q.worldguard.WorldGuard;
import com.sk89q.worldguard.protection.managers.RegionManager;
import com.sk89q.worldguard.protection.regions.ProtectedRegion;
import lombok.Getter;
import lombok.extern.slf4j.Slf4j;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.minimessage.tag.Tag;
import net.kyori.adventure.text.minimessage.tag.resolver.TagResolver;
import org.bukkit.*;
import org.bukkit.block.Block;
import org.bukkit.entity.Player;
import org.bukkit.event.Listener;
import org.bukkit.event.block.BlockBreakEvent;
import org.bukkit.inventory.ItemStack;
import org.bukkit.persistence.PersistentDataType;
import org.bukkit.scheduler.BukkitRunnable;
import org.bukkit.scheduler.BukkitTask;

import java.util.*;
import java.util.concurrent.ThreadLocalRandom;

import static ru.arc.util.TextUtil.mm;

@Slf4j
@Getter
public class Mine implements Listener {

    private final List<TemporaryBlock> tempBlocks = new ArrayList<>();
    private final TreeMap<Integer, Material> materialMap = new TreeMap<>();
    private final Set<Material> ores = new HashSet<>();
    private final int priority;
    ProtectedRegion region;
    List<Block> cache;
    BukkitTask replaceCobblestoneTask;
    BukkitTask replaceCache;
    int brokenBlocks = 0;
    int totalWeight;
    private final String permission;
    private final boolean particles;
    private final World world;
    private final String mineId;
    private final Material tempBlock;
    private final Material baseBlock;

    Config config = ConfigManager.of(ARC.plugin.getDataFolder().toPath(), "farms.yml");
    static long expireTime = 60000L;
    Map<UUID, Integer> blocksBrokenByPlayer = new HashMap<>();
    int maxBlocks;

    public Mine(String mineId, Map<Material, Integer> materialMap, String regionName, String worldName,
                Material tempBlock, int priority, Material baseBlock, String permission, boolean particles, int maxBlocks) {
        this.mineId = mineId;
        this.tempBlock = tempBlock;
        this.priority = priority;
        this.baseBlock = baseBlock;
        this.permission = permission;
        this.particles = particles;
        this.maxBlocks = maxBlocks;
        expireTime = config.integer("mine-config.expire-time", 60000);

        int totalWeight = 0;
        for (var entry : materialMap.entrySet()) {
            totalWeight += entry.getValue();
            this.materialMap.put(totalWeight, entry.getKey());
            ores.add(entry.getKey());
        }
        this.totalWeight = totalWeight;

        this.world = Bukkit.getWorld(worldName);
        if (world == null) {
            System.out.println(worldName + " world does not exist!");
            return;
        }

        RegionManager regionManager = WorldGuard.getInstance().getPlatform().getRegionContainer().get(BukkitAdapter.adapt(world));
        if (regionManager == null) {
            log.error("Region manager is null!");
            return;
        }
        this.region = regionManager.getRegion(regionName);

        if (region == null) {
            log.error("Region {} not found in world {}", regionName, worldName);
            return;
        }
        computeCache(true);
        setupTasks();
    }

    private Material pickRandomMaterial() {
        return materialMap.ceilingEntry((new Random()).nextInt(totalWeight)).getValue();
    }

    private void setupTasks() {
        cancelTasks();

        replaceCobblestoneTask = new BukkitRunnable() {
            @Override
            public void run() {
                List<TemporaryBlock> toRemove = new ArrayList<>();
                tempBlocks.stream()
                        .filter(TemporaryBlock::ifExpire)
                        .forEach(block -> {
                            CustomBlockData blockData = new CustomBlockData(block.block, ARC.plugin);
                            blockData.remove(new NamespacedKey(ARC.plugin, "t"));
                            block.block.setType(baseBlock);
                            toRemove.add(block);
                        });
                tempBlocks.removeAll(toRemove);
            }
        }.runTaskTimer(ARC.plugin, 20L, config.integer("mine-config.replace-time", 20));

        replaceCache = new BukkitRunnable() {

            @Override
            public void run() {
                if (brokenBlocks <= 0) return;

                Set<Integer> rng = new HashSet<>();
                for (int i = 0; i < config.integer("mine-config.replace-batch", 10); i++)
                    rng.add(ThreadLocalRandom.current().nextInt(cache.size()));

                for (int idx : rng) {
                    Block block = cache.get(idx);
                    if (block.getType() != baseBlock) continue;
                    Material material = pickRandomMaterial();
                    if (block.getRelative(0, -1, 0).getType() == Material.AIR) {
                        if (material == Material.SAND) material = Material.SANDSTONE;
                        else if (material == Material.GRAVEL) material = Material.STONE;
                    }
                    block.setType(material);
                    brokenBlocks--;
                    if (brokenBlocks <= 0) return;
                }
            }
        }.runTaskTimer(ARC.plugin, 25L, config.integer("mine-config.replace-time", 20));
    }

    private void computeCache(boolean replaceTemporaryBlock) {
        List<Block> blocks = new ArrayList<>();
        for (BlockVector3 vector3 : new CuboidRegion(BukkitAdapter.adapt(world), region.getMinimumPoint(), region.getMaximumPoint())) {
            int x = vector3.x();
            int y = vector3.y();
            int z = vector3.z();
            Block block = world.getBlockAt(x, y, z);

            if (block.getType() == tempBlock && replaceTemporaryBlock) {
                block.setType(baseBlock);
                CustomBlockData blockData = new CustomBlockData(block, ARC.plugin);
                blockData.remove(new NamespacedKey(ARC.plugin, "t"));
            }

            if ((world.getBlockAt(x + 1, y, z)).getType() == Material.AIR ||
                    (world.getBlockAt(x - 1, y, z)).getType() == Material.AIR ||
                    (world.getBlockAt(x, y + 1, z)).getType() == Material.AIR ||
                    (world.getBlockAt(x, y - 1, z)).getType() == Material.AIR ||
                    (world.getBlockAt(x, y, z + 1)).getType() == Material.AIR ||
                    (world.getBlockAt(x, y, z - 1)).getType() == Material.AIR)
                blocks.add(block);
        }

        this.cache = blocks;
    }

    public boolean processBreakEvent(BlockBreakEvent event) {
        Block block = event.getBlock();
        if (block.getLocation().getWorld() != world || region == null) return false;
        if (!region.contains(event.getBlock().getX(), event.getBlock().getY(), event.getBlock().getZ())) return false;
        if (event.getPlayer().hasPermission(config.string("admin-permission", "arc.farm-admin"))) return true;

        event.setCancelled(true);

        CustomBlockData blockData = new CustomBlockData(event.getBlock(), ARC.plugin);
        if (blockData.has(new NamespacedKey(ARC.plugin, "t"))) {
            Component text = config.componentDef("mine-config.already-broken", "<red>Этот блок еще не восстановился!");
            event.getPlayer().sendActionBar(text);
            return true;
        }

        if ((permission != null && !event.getPlayer().hasPermission(permission)) || !ores.contains(block.getType())) {
            event.getPlayer().sendMessage(TextUtil.noWGPermission());
            return true;
        }

        if (reachedMax(event.getPlayer())) {
            sendMaxReachedMessage(event.getPlayer());
            return true;
        } else if (block.getType() != baseBlock) incrementBlocks(event.getPlayer());

        Collection<ItemStack> stacks = event.getBlock().getDrops();
        stacks.forEach(stack -> event.getPlayer().getInventory().addItem(stack));

        if (block.getType() == baseBlock) event.getPlayer().giveExp(config.integer("mine-config.exp", 1));
        else event.getPlayer().giveExp(config.integer("mine-config.exp", 2));

        event.getBlock().setType(tempBlock);
        brokenBlocks++;

        blockData.set(new NamespacedKey(ARC.plugin, "t"), PersistentDataType.BOOLEAN, true);
        tempBlocks.add(new TemporaryBlock(event.getBlock()));

        Particle randomParticle = Utils.random(new Particle[]{Particle.FLAME, Particle.END_ROD, Particle.CRIT});
        if (particles) ParticleManager.queue(new ParticleBuilder(randomParticle)
                .receivers(List.of(event.getPlayer()))
                .location(block.getLocation().toCenterLocation())
                .count(config.integer("mine-config.particle-count", 5))
                .extra(config.real("mine-config.particle-extra", 0.06))
                .offset(config.real("mine-config.particle-offset", 0.25),
                        config.real("mine-config.particle-offset", 0.25),
                        config.real("mine-config.particle-offset", 0.25)));
        return true;
    }

    public void cancelTasks() {
        if (replaceCobblestoneTask != null && !replaceCobblestoneTask.isCancelled()) replaceCobblestoneTask.cancel();
        if (replaceCache != null && !replaceCache.isCancelled()) replaceCache.cancel();
    }

    private boolean reachedMax(Player player) {
        return blocksBrokenByPlayer.getOrDefault(player.getUniqueId(), 0) >= maxBlocks;
    }

    private void incrementBlocks(Player player) {
        blocksBrokenByPlayer.merge(player.getUniqueId(), 1, Integer::sum);
        int count = blocksBrokenByPlayer.get(player.getUniqueId());
        if (count % 16 == 0 && count != maxBlocks) {
            Component text = config.componentDef("mine-config.progress-message",
                    "<gray>Вы добыли <green><count><gray> из <gold><max><gray> за этот день",
                    TagResolver.builder()
                            .tag("count", Tag.inserting(mm(count + "")))
                            .tag("max", Tag.inserting(mm(maxBlocks + "")))
                            .build());
            player.sendActionBar(text);
        }
    }

    private void sendMaxReachedMessage(Player player) {
        if (CooldownManager.cooldown(player.getUniqueId(), "farm_limit_message") > 0) return;
        CooldownManager.addCooldown(player.getUniqueId(), "farm_limit_message", 60);

        Component text = config.componentDef("mine-config.limit-message", "<red>Вы достигли лимита добычи на сегодня</red>");
        player.sendMessage(text);
    }

    public void resetLimit() {
        blocksBrokenByPlayer.clear();
    }

    private static class TemporaryBlock {
        long timestamp;
        Block block;


        public TemporaryBlock(Block block) {
            this.block = block;
            this.timestamp = System.currentTimeMillis();
        }

        public boolean ifExpire() {
            return (System.currentTimeMillis() - timestamp > expireTime);
        }
    }

}

File: src/main/kotlin/ru/arc/farm/Mine.kt
package ru.arc.farm

import com.destroystokyo.paper.ParticleBuilder
import com.jeff_media.customblockdata.CustomBlockData
import net.kyori.adventure.text.minimessage.tag.Tag
import net.kyori.adventure.text.minimessage.tag.resolver.TagResolver
import org.bukkit.Material
import org.bukkit.NamespacedKey
import org.bukkit.Particle
import org.bukkit.block.Block
import org.bukkit.entity.Player
import org.bukkit.event.block.BlockBreakEvent
import org.bukkit.persistence.PersistentDataType
import org.bukkit.plugin.Plugin
import ru.arc.common.WeightedRandom
import ru.arc.core.ScheduledTask
import ru.arc.core.TaskScheduler
import ru.arc.util.CooldownManager
import ru.arc.util.ParticleManager
import ru.arc.util.RandomUtils
import ru.arc.util.TextUtil
import ru.arc.util.TextUtil.mm
import java.util.concurrent.ThreadLocalRandom

/**
 * Mine zone for breaking ores.
 *
 * Features:
 * - Weighted random ore generation
 * - Temporary blocks that regenerate after time
 * - Daily block limit per player
 * - Experience on mining
 * - Particles on break
 */
class Mine(
    override val id: String,
    override val priority: Int,
    private val config: MineConfig,
    private val region: FarmRegion,
    private val adminPermission: String,
    private val plugin: Plugin,
    private val scheduler: TaskScheduler,
    private val messages: FarmMessages = FarmMessages(),
    private val limitTracker: BlockLimitTracker = BlockLimitTracker(config.maxBlocksPerDay, 16),
    private val orePicker: WeightedRandom<Material> = createOrePicker(config.oreWeights),
    timeProvider: () -> Long = { System.currentTimeMillis() }
) : FarmZone {

    companion object {
        private fun createOrePicker(oreWeights: Map<Material, Int>): WeightedRandom<Material> {
            val picker = WeightedRandom<Material>()
            for ((material, weight) in oreWeights) {
                picker.add(material, weight.toDouble())
            }
            return picker
        }
    }

    private val tempBlockKey = NamespacedKey(plugin, "t")
    private val tempBlocks = TemporaryBlockTracker<Block>(config.expireTimeMs, timeProvider)
    private var brokenBlocks = 0
    private var blockCache: List<Block>? = null

    private var regenerateTask: ScheduledTask? = null
    private var respawnTask: ScheduledTask? = null

    /**
     * Start background tasks.
     */
    fun start() {
        computeBlockCache(replaceTemp = true)

        regenerateTask = scheduler.runTimer(20L, config.replaceTime) {
            processExpiredBlocks()
        }

        respawnTask = scheduler.runTimer(25L, config.replaceTime) {
            respawnOres()
        }
    }

    /**
     * Stop background tasks.
     */
    fun stop() {
        regenerateTask?.cancel()
        respawnTask?.cancel()
        regenerateTask = null
        respawnTask = null
    }

    override fun processBreak(event: BlockBreakEvent): BreakResult {
        if (!region.isValid()) return BreakResult.NotHandled

        val block = event.block
        if (!region.contains(block.location)) return BreakResult.NotHandled

        // Admin bypass
        if (event.player.hasPermission(adminPermission)) return BreakResult.Allowed

        event.isCancelled = true

        // Check if block is still regenerating
        val blockData = CustomBlockData(block, plugin)
        if (blockData.has(tempBlockKey)) {
            event.player.sendActionBar(mm(messages.alreadyBroken))
            return BreakResult.Cancelled
        }

        // Permission + material check
        if (!event.player.hasPermission(config.permission) || block.type !in config.ores) {
            event.player.sendMessage(TextUtil.noWGPermission())
            return BreakResult.Cancelled
        }

        // Limit check
        if (limitTracker.hasReachedLimit(event.player.uniqueId)) {
            sendLimitMessage(event.player)
            return BreakResult.Cancelled
        }

        // Only count towards limit if not base block
        if (block.type != config.baseBlock) {
            limitTracker.incrementBlocks(event.player.uniqueId)
        }

        // Give drops and exp
        mineBlock(event.player, block)

        if (limitTracker.shouldShowProgress(event.player.uniqueId)) {
            sendProgressMessage(event.player)
        }

        return BreakResult.Cancelled
    }

    private fun mineBlock(player: Player, block: Block) {
        // Give drops
        block.drops.forEach { player.inventory.addItem(it) }

        // Give exp
        val exp = if (block.type == config.baseBlock) config.expPerBase else config.expPerOre
        player.giveExp(exp)

        // Replace with temp block
        block.setType(config.tempBlock)
        brokenBlocks++

        // Mark as temp
        val blockData = CustomBlockData(block, plugin)
        blockData.set(tempBlockKey, PersistentDataType.BOOLEAN, true)
        tempBlocks.add(block)

        // Particles
        if (config.particles) {
            spawnParticles(player, block)
        }
    }

    private fun processExpiredBlocks() {
        val expired = tempBlocks.getExpired()
        for (block in expired) {
            val blockData = CustomBlockData(block, plugin)
            blockData.remove(tempBlockKey)
            block.setType(config.baseBlock)
        }
    }

    private fun respawnOres() {
        if (brokenBlocks <= 0) return

        val cache = blockCache ?: return
        if (cache.isEmpty()) return

        val indicesToCheck = (0 until config.replaceBatch)
            .map { ThreadLocalRandom.current().nextInt(cache.size) }
            .toSet()

        for (idx in indicesToCheck) {
            val block = cache[idx]
            if (block.type != config.baseBlock) continue

            var material = orePicker.random() ?: continue

            // Don't spawn falling blocks over air
            val below = block.getRelative(0, -1, 0)
            if (below.type == Material.AIR) {
                material = when (material) {
                    Material.SAND -> Material.SANDSTONE
                    Material.GRAVEL -> Material.STONE
                    else -> material
                }
            }

            block.setType(material)
            brokenBlocks--

            if (brokenBlocks <= 0) return
        }
    }

    private fun computeBlockCache(replaceTemp: Boolean) {
        val blocks = mutableListOf<Block>()

        for (block in region.getBlocks()) {
            // Replace temp blocks on startup
            if (replaceTemp && block.type == config.tempBlock) {
                block.setType(config.baseBlock)
                CustomBlockData(block, plugin).remove(tempBlockKey)
            }

            // Only cache blocks adjacent to air (visible)
            if (isAdjacentToAir(block)) {
                blocks.add(block)
            }
        }

        blockCache = blocks
    }

    private fun isAdjacentToAir(block: Block): Boolean {
        return block.getRelative(1, 0, 0).type == Material.AIR ||
            block.getRelative(-1, 0, 0).type == Material.AIR ||
            block.getRelative(0, 1, 0).type == Material.AIR ||
            block.getRelative(0, -1, 0).type == Material.AIR ||
            block.getRelative(0, 0, 1).type == Material.AIR ||
            block.getRelative(0, 0, -1).type == Material.AIR
    }

    private fun spawnParticles(player: Player, block: Block) {
        val particle = RandomUtils.random(arrayOf(Particle.FLAME, Particle.END_ROD, Particle.CRIT))
        ParticleManager.queue(
            ParticleBuilder(particle)
                .receivers(listOf(player))
                .location(block.location.toCenterLocation())
                .count(5)
                .extra(0.06)
                .offset(0.25, 0.25, 0.25)
        )
    }

    private fun sendLimitMessage(player: Player) {
        if (CooldownManager.cooldown(player.uniqueId, "farm_limit_message") > 0) return
        CooldownManager.addCooldown(player.uniqueId, "farm_limit_message", messages.limitMessageCooldown.toLong())
        player.sendMessage(mm(messages.limitReached))
    }

    private fun sendProgressMessage(player: Player) {
        val count = limitTracker.getBlockCount(player.uniqueId)
        val text = mm(
            messages.progress,
            TagResolver.builder()
                .tag("count", Tag.inserting(mm(count.toString())))
                .tag("max", Tag.inserting(mm(config.maxBlocksPerDay.toString())))
                .build()
        )
        player.sendActionBar(text)
    }

    override fun resetLimits() {
        limitTracker.resetAll()
    }

    override fun cleanup() {
        stop()
    }
}

